File: src/mihomo_manager.rs
use std::sync::Arc;

use tokio::process::Command;
use tokio::sync::RwLock;
use tokio::time::{self, Duration};

use crate::config::MihomoConfig;

#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum MihomoStatus {
    Running,
    Stopped,
    Starting,
    Crashed,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct MihomoState {
    pub status: MihomoStatus,
    pub pid: Option<u32>,
    pub restarts: u32,
    pub last_error: Option<String>,
}

impl Default for MihomoState {
    fn default() -> Self {
        Self {
            status: MihomoStatus::Stopped,
            pid: None,
            restarts: 0,
            last_error: None,
        }
    }
}

#[derive(Clone)]
pub struct MihomoManager {
    pub state: Arc<RwLock<MihomoState>>,
    mihomo_config: MihomoConfig,
}

impl MihomoManager {
    pub fn new(mihomo_config: MihomoConfig) -> Self {
        Self {
            state: Arc::new(RwLock::new(MihomoState::default())),
            mihomo_config,
        }
    }

    /// Check if the mihomo binary exists in PATH.
    pub fn find_binary() -> Option<String> {
        let output = std::process::Command::new("which")
            .arg("mihomo")
            .output()
            .ok()?;
        if output.status.success() {
            let path = String::from_utf8_lossy(&output.stdout).trim().to_string();
            if path.is_empty() {
                None
            } else {
                Some(path)
            }
        } else {
            None
        }
    }

    /// Check if mihomo API is reachable.
    async fn is_api_alive(&self) -> bool {
        let url = format!("{}/version", self.mihomo_config.api.trim_end_matches('/'));
        let mut req = reqwest::Client::new()
            .get(&url)
            .timeout(Duration::from_secs(2));
        if let Some(secret) = &self.mihomo_config.api_secret {
            req = req.header("Authorization", format!("Bearer {secret}"));
        }
        matches!(req.send().await, Ok(resp) if resp.status().is_success())
    }

    /// Start mihomo process and return the Child.
    async fn start_process(&self) -> Result<tokio::process::Child, String> {
        let output_path = self.mihomo_config.output.display().to_string();

        let child = Command::new("mihomo")
            .arg("-f")
            .arg(&output_path)
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::piped())
            .kill_on_drop(true)
            .spawn()
            .map_err(|e| format!("failed to spawn mihomo: {e}"))?;

        Ok(child)
    }

    /// Main lifecycle loop: start mihomo, monitor, restart on crash.
    pub async fn run(self) {
        const HEALTH_INTERVAL: Duration = Duration::from_secs(10);
        const STARTUP_GRACE: Duration = Duration::from_secs(3);
        const RESTART_BACKOFF: Duration = Duration::from_secs(5);

        loop {
            // Check if already running externally
            if self.is_api_alive().await {
                tracing::info!("mihomo API already reachable, attaching to existing process");
                {
                    let mut s = self.state.write().await;
                    s.status = MihomoStatus::Running;
                    s.last_error = None;
                }
                self.health_loop(HEALTH_INTERVAL).await;
                // health_loop returns when API becomes unreachable
                tracing::warn!("mihomo API lost, will attempt restart");
            }

            // Start mihomo
            {
                let mut s = self.state.write().await;
                s.status = MihomoStatus::Starting;
            }

            let child = match self.start_process().await {
                Ok(child) => child,
                Err(e) => {
                    tracing::error!(error = %e, "failed to start mihomo");
                    let mut s = self.state.write().await;
                    s.status = MihomoStatus::Crashed;
                    s.last_error = Some(e);
                    time::sleep(RESTART_BACKOFF).await;
                    continue;
                }
            };

            let pid = child.id();
            tracing::info!(pid = pid, "mihomo process started");

            {
                let mut s = self.state.write().await;
                s.pid = pid;
                s.status = MihomoStatus::Starting;
            }

            // Wait for API to come up
            time::sleep(STARTUP_GRACE).await;

            let api_up = {
                let mut up = false;
                for _ in 0..5 {
                    if self.is_api_alive().await {
                        up = true;
                        break;
                    }
                    time::sleep(Duration::from_secs(1)).await;
                }
                up
            };

            if api_up {
                tracing::info!("mihomo API is up");
                {
                    let mut s = self.state.write().await;
                    s.status = MihomoStatus::Running;
                    s.last_error = None;
                }
            } else {
                tracing::warn!("mihomo started but API not responding");
                {
                    let mut s = self.state.write().await;
                    s.status = MihomoStatus::Running;
                    s.last_error = Some("API not responding after start".into());
                }
            }

            // Monitor: wait for process exit OR API health failure
            self.monitor_loop(child, HEALTH_INTERVAL).await;

            // Process died or API unreachable
            {
                let mut s = self.state.write().await;
                s.status = MihomoStatus::Crashed;
                s.pid = None;
                s.restarts += 1;
                tracing::warn!(restarts = s.restarts, "mihomo exited, restarting...");
            }

            time::sleep(RESTART_BACKOFF).await;
        }
    }

    /// Poll API until it becomes unreachable. Returns when health check fails.
    async fn health_loop(&self, interval: Duration) {
        let mut ticker = time::interval(interval);
        let mut consecutive_failures = 0u32;
        loop {
            ticker.tick().await;
            if self.is_api_alive().await {
                consecutive_failures = 0;
            } else {
                consecutive_failures += 1;
                tracing::warn!(failures = consecutive_failures, "mihomo health check failed");
                if consecutive_failures >= 3 {
                    let mut s = self.state.write().await;
                    s.status = MihomoStatus::Crashed;
                    s.last_error = Some("API unreachable (3 consecutive failures)".into());
                    return;
                }
            }
        }
    }

    /// Monitor a child process we spawned. Returns when the process exits
    /// or API health checks fail 3 times consecutively.
    async fn monitor_loop(&self, mut child: tokio::process::Child, interval: Duration) {
        let mut ticker = time::interval(interval);
        let mut consecutive_failures = 0u32;
        loop {
            tokio::select! {
                status = child.wait() => {
                    match status {
                        Ok(exit) => {
                            let mut s = self.state.write().await;
                            s.last_error = Some(format!("process exited: {exit}"));
                        }
                        Err(e) => {
                            let mut s = self.state.write().await;
                            s.last_error = Some(format!("wait error: {e}"));
                        }
                    }
                    return;
                }
                _ = ticker.tick() => {
                    if self.is_api_alive().await {
                        consecutive_failures = 0;
                    } else {
                        consecutive_failures += 1;
                        tracing::warn!(failures = consecutive_failures, "mihomo health check failed");
                        if consecutive_failures >= 3 {
                            tracing::error!("mihomo unresponsive, killing process");
                            let _ = child.kill().await;
                            let mut s = self.state.write().await;
                            s.last_error = Some("killed: API unreachable".into());
                            return;
                        }
                    }
                }
            }
        }
    }
}

File: zig/src/mihomo_manager.zig
const std = @import("std");
const config = @import("config.zig");
const model = @import("model.zig");
const state = @import("state.zig");
const mihomo_api = @import("mihomo_api.zig");

pub const MihomoManager = struct {
    gpa: std.mem.Allocator,
    io: std.Io,
    shared_state: *model.MihomoState,
    lock: *state.ThreadRwLock,
    mihomo_config: config.MihomoConfig,

    pub fn init(io: std.Io, gpa: std.mem.Allocator, mcfg: config.MihomoConfig, shared_state: *model.MihomoState, lock: *state.ThreadRwLock) MihomoManager {
        return .{
            .gpa = gpa,
            .io = io,
            .shared_state = shared_state,
            .lock = lock,
            .mihomo_config = mcfg,
        };
    }

    fn updateState(self: *MihomoManager, s: model.MihomoStatus, pid: ?u32, last_error: ?[]const u8, inc_restarts: bool) void {
        self.lock.lock();
        defer self.lock.unlock();
        if (self.shared_state.last_error) |e| self.gpa.free(e);
        self.shared_state.status = s;
        if (pid) |p| self.shared_state.pid = p;
        self.shared_state.last_error = last_error;
        if (inc_restarts) self.shared_state.restarts += 1;
    }

    fn clearPid(self: *MihomoManager) void {
        self.lock.lock();
        defer self.lock.unlock();
        self.shared_state.pid = null;
    }

    pub fn run(self: *MihomoManager) void {
        const io = self.io;
        const HEALTH_INTERVAL: u64 = 10 * std.time.ns_per_s;

        while (true) {
            if (self.isApiAlive()) {
                std.log.info("mihomo API already reachable", .{});
                self.updateState(.running, null, null, false);
                self.healthLoop(HEALTH_INTERVAL);
                std.log.warn("mihomo API lost, will attempt restart", .{});
            }

            self.updateState(.starting, null, null, false);
            var child = self.startProcess() catch |err| {
                std.log.err("failed to start mihomo: {s}", .{@errorName(err)});
                self.updateState(.crashed, null, std.fmt.allocPrint(self.gpa, "{s}", .{@errorName(err)}) catch null, false);
                io.sleep(std.Io.Duration.fromSeconds(5), .real) catch break;
                continue;
            };

            const pid: u32 = @intCast(child.id orelse 0);
            std.log.info("mihomo process started pid={d}", .{pid});
            self.updateState(.starting, pid, null, false);

            // Wait for API to come up
            io.sleep(std.Io.Duration.fromSeconds(3), .real) catch break;
            var api_up = false;
            var retry: u32 = 0;
            while (retry < 5) : (retry += 1) {
                if (self.isApiAlive()) {
                    api_up = true;
                    break;
                }
                io.sleep(std.Io.Duration.fromSeconds(1), .real) catch break;
            }

            if (api_up) {
                std.log.info("mihomo API is up", .{});
                self.updateState(.running, null, null, false);
            } else {
                std.log.warn("mihomo started but API not responding", .{});
                self.updateState(.running, null, std.fmt.allocPrint(self.gpa, "API not responding after start", .{}) catch null, false);
            }

            // Monitor: block until child exits
            self.monitorLoop(&child, HEALTH_INTERVAL);

            self.updateState(.crashed, null, null, true);
            self.clearPid();
            std.log.warn("mihomo exited, restarts={d}", .{self.shared_state.restarts});
            io.sleep(std.Io.Duration.fromSeconds(5), .real) catch break;
        }
    }

    fn isApiAlive(self: *MihomoManager) bool {
        const url = std.mem.trimEnd(u8, self.mihomo_config.api, "/");
        const full = std.fmt.allocPrint(self.gpa, "{s}/version", .{url}) catch return false;
        defer self.gpa.free(full);

        var client: std.http.Client = .{ .allocator = self.gpa, .io = self.io };
        defer client.deinit();

        var req = client.request(.GET, std.Uri.parse(full) catch return false, .{}) catch return false;
        var buf: [256]u8 = undefined;
        if (self.mihomo_config.api_secret) |secret| {
            const auth = std.fmt.bufPrint(&buf, "Bearer {s}", .{secret}) catch return false;
            req.headers.authorization = .{ .override = auth };
        }
        defer req.deinit();

        req.sendBodiless() catch return false;
        var response = req.receiveHead(&.{}) catch return false;
        return response.head.status.class() == .success;
    }

    fn startProcess(self: *MihomoManager) !std.process.Child {
        return std.process.spawn(self.io, .{
            .argv = &.{ "mihomo", "-f", self.mihomo_config.output },
            .stdout = .ignore,
            .stderr = .ignore,
        });
    }

    fn monitorLoop(self: *MihomoManager, child: *std.process.Child, health_interval_ns: u64) void {
        const io = self.io;
        var failures: u32 = 0;
        _ = health_interval_ns;

        while (true) {
            {
                if (self.isApiAlive()) {
                    failures = 0;
                } else {
                    failures += 1;
                    std.log.warn("mihomo health check failed failures={d}", .{failures});
                    if (failures >= 3) {
                        std.log.err("mihomo unresponsive, killing process", .{});
                        child.kill(io);
                        self.updateState(.crashed, null, std.fmt.allocPrint(self.gpa, "killed: API unreachable", .{}) catch null, false);
                        return;
                    }
                }
            }

            // Check if child exited (non-blocking poll)
            const term = child.wait(io) catch {
                io.sleep(std.Io.Duration.fromSeconds(1), .real) catch return;
                continue;
            };
            self.updateState(.crashed, null, std.fmt.allocPrint(self.gpa, "process exited: {}", .{term}) catch null, false);
            return;
        }
    }

    fn healthLoop(self: *MihomoManager, interval_ns: u64) void {
        const io = self.io;
        var failures: u32 = 0;
        while (true) {
            io.sleep(std.Io.Duration.fromNanoseconds(@intCast(interval_ns)), .real) catch return;
            if (self.isApiAlive()) {
                failures = 0;
            } else {
                failures += 1;
                std.log.warn("mihomo health check failed failures={d}", .{failures});
                if (failures >= 3) {
                    self.updateState(.crashed, null, std.fmt.allocPrint(self.gpa, "API unreachable (3 failures)", .{}) catch null, false);
                    return;
                }
            }
        }
    }
};
